packages/utils/src/peripherals/relayer.ts

```diff
@@ -0,0 +1,114 @@
+import { getChainIdFromDomain, getDecimalsForAsset } from "../helpers";
+import { Logger } from "../logging";
+
+import { ChainData } from "./chainData";
+import { getConversionRate, getGelatoEstimatedFee } from "./gelato";
+
+const relayerBufferPercentage = 20; // 20% bump on total estimated relayer fee
+
+export const calculateRelayerFee = async (
+  params: {
+    originDomain: string;
+    destinationDomain: string;
+    originNativeToken?: string;
+    destinationNativeToken?: string;
+    callDataGasAmount?: number;
+    isHighPriority?: boolean;
+  },
+  chainData: Map<string, ChainData>,
+  logger?: Logger,
+): Promise<BigNumber> => {
+  const {
+    originDomain,
+    destinationDomain,
+    callDataGasAmount,
+    originNativeToken: _originNativeToken,
+    destinationNativeToken: _destinationNativeToken,
+    isHighPriority: _isHighPriority,
+  } = params;
+
+  const originNativeToken = _originNativeToken ?? constants.AddressZero;
+  const destinationNativeToken = _destinationNativeToken ?? constants.AddressZero;
+  const isHighPriority = _isHighPriority ?? false;
+
+  const originChainId = await getChainIdFromDomain(originDomain, chainData);
+  const destinationChainId = await getChainIdFromDomain(destinationDomain, chainData);
+
+  // fetch executeGasAmount from chainData
+  const {
+    execute: executeGasAmount,
+    executeL1: executeL1GasAmount,
+    gasPriceFactor,
+  } = await getHardcodedGasLimits(originChainId, chainData);
+  if (logger) {
+    logger.debug("Hardcoded gasLimits", undefined, undefined, {
+      execute: executeGasAmount,
+      executeL1: executeL1GasAmount,
+      gasPriceFactor,
+    });
+  }
+
+  const totalGasAmount = callDataGasAmount
+    ? Number(executeGasAmount) + Number(callDataGasAmount)
+    : Number(executeGasAmount);
+  const estimatedRelayerFee = await getGelatoEstimatedFee(
+    destinationChainId,
+    destinationNativeToken,
+    Number(totalGasAmount),
+    isHighPriority,
+  );
+
+  if (logger) {
+    logger.info("Gas Price estimates", undefined, undefined, {
+      originNativeToken,
+      originChainId,
+      destinationNativeToken,
+      destinationChainId,
+      executeGasAmount,
+      callDataGasAmount,
+    });
+  }
+
+  // add relayerFee bump to estimatedRelayerFee
+  const bumpedFee = estimatedRelayerFee.add(estimatedRelayerFee.mul(BigNumber.from(relayerBufferPercentage)).div(100));
+
+  // TODO: Convert the estimatedRelayerFee to the originNativeToken
+  const [originTokenPrice, destinationTokenPrice, originTokenDecimals, destinationTokenDecimals] = await Promise.all([
+    getConversionRate(originChainId, undefined, undefined),
+    getConversionRate(destinationChainId, undefined, undefined),
+    getDecimalsForAsset(originNativeToken, originChainId, undefined, chainData),
+    getDecimalsForAsset(destinationNativeToken, destinationChainId, undefined, chainData),
+  ]);
+
+  if (originTokenPrice == 0 || destinationTokenPrice == 0) {
```

Comment:
What if the origin token has a price of 0 but the destination has a price of > 0? I can't imagine this ever being the case but... idk, maybe we should throw an error? Because the destination is where we're ultimately paying for the tx...

Weird edge case, but worth accounting for. I think throw an error here if `originTokenPrice == 0 && destinationTokenPrice > 0`

if origin native token is 0, that would mean user would have no way to pay fees using a worthless native token... but I'm more worried about the edge case where the resource fails (i.e. `getConversionRate`'s pricing is incorrect)

Alternatively if `originTokenPrice > 0 && destinationTokenPrice == 0` we actually don't care about paying fees at all - it's free on the destination! So we should return `BigNumber.from(0)` like we're already doing in that case.

Reply:
getConversionRate is the gelato utility function which was getting used in nxtp v1. This is a util package so I guess its better to return `0` and process it on agents rather than throwing an error.